lib/usagereporter/teleport/aggregating/reporter.go

```diff
@@ -0,0 +1,316 @@
+// will be used for all backend operations.
+func NewReporter(
+	ctx context.Context,
+	cfg ReporterConfig,
+) (*Reporter, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	anonymizer, err := utils.NewHMACAnonymizer(cfg.ClusterName.GetClusterID())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	baseCtx, baseCancel := context.WithCancel(ctx)
+	periodicCtx, periodicCancel := context.WithCancel(baseCtx)
+
+	r := &Reporter{
+		anonymizer: anonymizer,
+		svc:        reportService{cfg.Backend},
+		log:        cfg.Log,
+
+		clusterName:    anonymizer.AnonymizeNonEmpty(cfg.ClusterName.GetClusterName()),
+		reporterHostID: anonymizer.AnonymizeNonEmpty(cfg.ReporterHostID),
+
+		baseCtx:        baseCtx,
+		baseCancel:     baseCancel,
+		periodicCancel: periodicCancel,
+	}
+
+	r.wg.Add(1)
+	go r.runPeriodicFinalize(periodicCtx)
+
+	return r, nil
+}
+
+// Reporter aggregates and persists usage events to the backend.
+type Reporter struct {
+	anonymizer utils.Anonymizer
+	svc        reportService
+	log        logrus.FieldLogger
+
+	// clusterName is the anonymized cluster name.
+	clusterName []byte
+	// reporterHostID is the anonymized host ID of the reporter (this agent).
+	reporterHostID []byte
+
+	// wg tracks all long-lived or I/O bound goroutines, to be waited on during
+	// GracefulStop.
+	wg sync.WaitGroup
+	// baseCtx is the parent context for all operations (backend operations
+	// being the most important).
+	baseCtx context.Context
+	// baseCancel cancels baseCtx.
+	baseCancel context.CancelFunc
+	// periodicCancel cancels the context used by the periodic finalizer; child
+	// of baseCtx.
+	periodicCancel context.CancelFunc
+
+	// mu protects stopped, startTime and userActivity.
+	mu sync.Mutex
+	// stopped indicates that startTime and userActivity should not be touched
+	// anymore, that all events should be dropped, and that no goroutines
+	// tracked by wg should be started.
+	stopped bool
+
+	// startTime is the start time of the currently active data collection.
+	startTime time.Time
+	// userActivity is a map of username (non-anonymized) to UserActivityRecord.
+	userActivity map[string]*prehogv1.UserActivityRecord
+}
+
+var _ usagereporter.GracefulStopper = (*Reporter)(nil)
+
+// GracefulStop implements [usagereporter.GracefulStopper]. Can only be called
+// at most once.
+func (r *Reporter) GracefulStop(ctx context.Context) error {
+	r.periodicCancel()
+
+	r.mu.Lock()
+
+	r.stopped = true
+
+	startTime := r.startTime
+	userActivity := r.userActivity
+	// release memory
+	r.userActivity = nil
+
+	r.mu.Unlock()
+
+	if len(userActivity) > 0 {
+		// we're allowed to do this because we know that we'll only Wait on r.wg
+		// later on in this function's body; anywhere else we're not allowed to
+		// Add while r.stopped is true, as Wait can't be called concurrently
+		// with an Add from 0, and if we began shutting down we can't assume
+		// that the count is above 0
+		r.wg.Add(1)
```

Comment:
We call `Add` in three points: right before the periodic finalizer, and in the two places right before we spawn `persistUserActivity`; there's two deferred `Done`s (in `runPeriodicFinalize` and `persistUserActivity`) and a single `Wait`.

We can simplify this further by not caring if there's leftover goroutines still running before returning from `GracefulStop` I suppose, since they'll just be stopping due to context closure anyway, but is that actually an improvement?

Reply:
edit: we actually can't, we rely on the waitgroup to confirm the end of the goroutines to actually give enough time for the backend operations to finish, letting them die on their own only works if the shutdown context is actually too short